src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs

```diff
@@ -1698,19 +1702,66 @@ internal HashSet<NamedTypeSymbol> GetForwardedTypes()
                 return ((CommonAssemblyWellKnownAttributeData)attributesBag.DecodedWellKnownAttributeData)?.ForwardedTypes;
             }
 
-            attributesBag = null;
-            LoadAndValidateAttributes(OneOrMany.Create(GetAttributeDeclarations()), ref attributesBag, attributeMatchesOpt: this.IsPossibleForwardedTypesAttribute);
+            var allocate = t_forwardedTypesAttributesInProgress is null;
+
+            if (allocate)
+            {
+                t_forwardedTypesAttributesInProgress = PooledHashSet<AttributeSyntax>.GetInstance();
+            }
+
+            try
+            {
+                attributesBag = null;
+                LoadAndValidateAttributes(
+                    OneOrMany.Create(GetAttributeDeclarations()), ref attributesBag,
+                    attributeMatchesOpt: this.IsPossibleForwardedTypesAttribute,
+                    beforeAttributePartBound: BeforePossibleForwardedTypesAttributePartBound,
+                    afterAttributePartBound: AfterPossibleForwardedTypesAttributePartBound);
```

Comment:
How often is `GetForwardedTypes` called? This introduces two delegate allocations. If it's a hot path think we should consider caching them out but if not think we can avoid complicating the code here.

Reply:
> How often is GetForwardedTypes called? This introduces two delegate allocations. If it's a hot path think we should consider caching them out but if not think we can avoid complicating the code here. 

I was following what we did around the delegate based on ```IsPossibleForwardedTypesAttribute``` above. I.e. we were not trying to cache it, even though we could. Also, once we merge the support for caching delegates targeting static methods (which I hope going to happen soon), these will be cached.

Reply:
@jaredpar Would you like me to add manual delegate caching for the two ne delegates here?

Reply:
I think we can avoid doing it for now. If it was a known hot path I would say cache but if it's not then we can cache later if data reveals its a rpoblem.